src/Pool.sol

```diff
@@ -213,51 +216,93 @@ abstract contract Pool {
     }
 
     /// @notice Removes from the history and returns the amount of received
     /// funds available for collection by the user
     /// @param receiverAddr The address of the receiver
     /// @return collected The collected amount
-    function _collectInternal(address receiverAddr) internal returns (uint128 collected) {
+    function _collectInternal(address receiverAddr)
+        internal
+        returns (uint128 collected, uint128 dripped)
+    {
         Receiver storage receiver = receivers[receiverAddr];
+
+        // Collectable independently from cycles
+        collected = receiver.collectable;
+        if (collected > 0) receiver.collectable = 0;
+
+        // Collectable from cycles
         uint64 collectedCycle = receiver.nextCollectedCycle;
-        if (collectedCycle == 0) return 0;
         uint64 currFinishedCycle = _currTimestamp() / cycleSecs;
-        if (collectedCycle > currFinishedCycle) return 0;
-        int128 lastFundsPerCycle = receiver.lastFundsPerCycle;
-        for (; collectedCycle <= currFinishedCycle; collectedCycle++) {
-            lastFundsPerCycle += receiver.amtDeltas[collectedCycle - 1].nextCycle;
-            lastFundsPerCycle += receiver.amtDeltas[collectedCycle].thisCycle;
-            collected += uint128(lastFundsPerCycle);
-            delete receiver.amtDeltas[collectedCycle - 1];
+        if (collectedCycle != 0 && collectedCycle <= currFinishedCycle) {
+            int128 lastFundsPerCycle = int128(receiver.lastFundsPerCycle);
+            for (; collectedCycle <= currFinishedCycle; collectedCycle++) {
+                lastFundsPerCycle += receiver.amtDeltas[collectedCycle - 1].nextCycle;
+                delete receiver.amtDeltas[collectedCycle - 1];
+                lastFundsPerCycle += receiver.amtDeltas[collectedCycle].thisCycle;
+                collected += uint128(lastFundsPerCycle);
+            }
+            receiver.lastFundsPerCycle = uint128(lastFundsPerCycle);
+            receiver.nextCollectedCycle = collectedCycle;
+        }
+
+        // Dripped when collected
+        Sender storage sender = senders[receiverAddr];
+        if (collected > 0 && sender.dripsFraction > 0 && sender.weightSum > 0) {
+            dripped = (collected * sender.dripsFraction) / DRIPS_FRACTION_MAX;
+            dripped -= dripped % sender.weightSum;
+            collected -= dripped;
+            uint128 drippedPerWeight = dripped / sender.weightSum;
+            // Iterating over receivers, see `ReceiverWeights` for details
+            address dripsAddr = ReceiverWeightsImpl.ADDR_ROOT;
+            address hint = ReceiverWeightsImpl.ADDR_ROOT;
+            while (true) {
+                uint32 weight;
+                (dripsAddr, hint, weight) = sender.receiverWeights.nextWeight(dripsAddr, hint);
+                if (dripsAddr == ReceiverWeightsImpl.ADDR_ROOT) break;
+                receivers[dripsAddr].collectable += drippedPerWeight * weight;
```

Comment:
Should  we add some sanity checks? I already checked `weightSum` seems correct, etc.

We could track the `totalDripped` here in the loop and afterwards 
```
require(dripped == totalDdripped)
```

At least for stress testing it, would be maybe nice to have it.

Reply:
I'm a bit skeptical about usefulness of adding `require(dripped == totalDripped)` to production code for sake of stress testing. We must test if everybody actually gets what they're supposed to get and if `dripped != totalDripped`, these checks will catch that and many more issues.

Reply:
Added the sanity check

Reply:
Okay nice. However, if we run into a problem and it reverts the fund would be locked.